components/supervisor/pkg/supervisor/services.go

```diff
@@ -1033,3 +1035,149 @@ func (s *portService) RetryAutoExpose(ctx context.Context, req *api.RetryAutoExp
 func (s *statusService) ResourcesStatus(ctx context.Context, in *api.ResourcesStatuRequest) (*api.ResourcesStatusResponse, error) {
 	return s.topService.data, nil
 }
+
+type taskService struct {
+	tasksManager *tasksManager
+
+	api.UnimplementedTaskServiceServer
+}
+
+func (s *taskService) RegisterGRPC(srv *grpc.Server) {
+	api.RegisterTaskServiceServer(srv, s)
+}
+
+func (s *taskService) RegisterREST(mux *runtime.ServeMux, grpcEndpoint string) error {
+	return api.RegisterPortServiceHandlerFromEndpoint(context.Background(), mux, grpcEndpoint, []grpc.DialOption{grpc.WithTransportCredentials(insecure.NewCredentials())})
+}
+
+// ListenToOutput listens to the output of a task. It streams the output from the task's file and ends when the task's state changes to done.
+func (s *taskService) ListenToOutput(req *api.ListenToOutputRequest, srv api.TaskService_ListenToOutputServer) error {
+	fileLocation := logs.PrebuildLogFileName(logs.TerminalStoreLocation, req.TaskId)
+	if _, err := os.Stat(fileLocation); os.IsNotExist(err) {
+		return status.Error(codes.NotFound, "task not found")
+	}
+
+	file, err := os.Open(fileLocation)
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
+	defer file.Close()
+
+	buf := make([]byte, 4096)
+
+	// If the task is already done, we can just send all the logs and return without subscribing to task updates
+	taskStatuses := s.tasksManager.Status()
+	taskClosed := false
+	for _, taskStatus := range taskStatuses {
+		if taskStatus.Id == req.TaskId && taskStatus.State == api.TaskState_closed {
+			taskClosed = true
+			break
+		}
+	}
+
+	if taskClosed {
+		buf := make([]byte, 4096)
+		for {
+			n, err := file.Read(buf)
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				return status.Error(codes.Internal, err.Error())
+			}
+
+			if err := srv.Send(&api.ListenToOutputResponse{Data: buf[:n]}); err != nil {
+				return status.Error(codes.Internal, err.Error())
+			}
+		}
+
+		return nil
+	}
+
+	done := make(chan struct{})
+
+	sub := s.tasksManager.Subscribe()
+	if sub == nil {
+		log.Warn("potentially leaking subscription to tasks status: too many subscriptions")
+		return status.Error(codes.ResourceExhausted, "too many subscriptions")
+	}
+	defer sub.Close()
+
+	go func() {
+		defer close(done)
+		for {
+			select {
+			case <-srv.Context().Done():
+				return
+			case updates := <-sub.Updates():
+				if updates == nil {
+					return
+				}
+				for _, update := range updates {
+					if update == nil {
+						continue
+					}
+					if update.Id == req.TaskId && update.State == api.TaskState_closed {
```

Comment:
I think what would help the structure of this method:
 1. unconditionally resolve the "closed" state of the task (uses getTaskStatus & the `subscribe`, output: `taskClosedChan`)
 2. (optional) setup the fs watcher
 3. start a single loop that reads from the file
   1. in the EOF case, stop if `isClosed == true`, continue otherwise
   3. stop in case of ctx.Done()
   4. use a ticker (1s) to read from the fs again (optional: use the channel of the fs watcher instead)

Reply:
> Happy to talk this through if you'd like to.

@filiptronicek Awesome, sent an invite for tomorrow morning. Please move to better fit your schedule! :+1: